### DISCUSSION TITLE
add a super class or interface for all prisma model ?

### DISCUSSION THREAD

**Discussion by superiums** [prisma/prisma#14230] [Q&A]
## Problem

i'm using some crud generater for prisma in nest.  and i found it's hard to abstract a super class for all prisma model.

for example: 
i has prismaClient.user and prismaClient.post
i want to write a common service for all services. 

```
class userService extends commonService<User>:{
    constructor(){
        super.model=user
    }
    ...
}
```
commonService
```
class commonService<T extends SOME_CLASS_SUPER_OF_PRISMA_MODEL>:{
    model : T

    create(): Promise<T>{
        T.create()    // instead of prisma.user.create() 
        ...
    }
}
```

## Suggested solution

all model actions like create/findOne/delete should extends some super class.(or implents some interface).
this makes all model more united and structured.

**Comment by austincrim** [comment]
Hey @superiums 👋,

I would take a look through [this existing thread](https://github.com/prisma/prisma/issues/5273) about doing something similar to this. The latest ideas and workarounds will be in there 👍.

**Comment by cyrilchapon** [comment]
+1, exact same need with same use-case (Nest "SuperModel" creation, with common operations).

To expand on the discussion, here's the goal I'm tryin to achieve which might be similar to @superiums :

```ts
import { Prisma, PrismaClient } from '@prisma/client'
import { PrismaService } from '../prisma/prisma.service'

export const defaultListAllowedParams = [
  'skip',
  'take',
  'cursor',
  'where',
  'orderBy',
] as const

export type DefaultListAllowedParams = typeof defaultListAllowedParams[number]
type ListWithAllowedParams = {
  [key in DefaultListAllowedParams]?: any
}

export type PickListAllowedParams<T extends ListWithAllowedParams> = Pick<
  T,
  DefaultListAllowedParams
>

export const defaultPrimaryKey = 'id'
export const defaultTimestamps = ['createdAt', 'updatedAt'] as const
export const defaultHiddenFields = [
  defaultPrimaryKey,
  ...defaultTimestamps,
] as const

export type DefaultHiddenFields = typeof defaultHiddenFields[number]

export type OmitHiddenFields<T> = Omit<T, DefaultHiddenFields>

export class BaseModel<
  ModelDelegateName,
  Model,
  WhereUniqueInput,
  FindManyArgs,
  CreateInput,
  UpdateInput
> {
  constructor(
    private prisma: PrismaService,
    private modelDelegateName: ModelDelegateName
  ) {}

  async find(
    whereUniqueInput: WhereUniqueInput,
  ): Promise<Model | null> {
    return this.prisma[this.modelDelegateName].findUnique({
      where: whereUniqueInput,
    })
  }

  async list(
    findManyArgs: PickListAllowedParams<FindManyArgs>,
  ): Promise<Model[]> {
    return this.prisma[this.modelDelegateName].findMany(findManyArgs)
  }

  async create(data: OmitHiddenFields<CreateInput>): Promise<Model> {
    return this.prisma[this.modelDelegateName].create({
      data,
    })
  }

  async update(params: {
    where: WhereUniqueInput
    data: OmitHiddenFields<UpdateInput>
  }): Promise<Model> {
    const { data, where } = params
    return this.prisma[this.modelDelegateName].update({
      data,
      where,
    })
  }

  async delete(where: WhereUniqueInput): Promise<Model> {
    return this.prisma[this.modelDelegateName].delete({
      where,
    })
  }
}

```

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
